cmake: Toolchain abstraction

mped-oticon opened the pull request:
This PR:
 * Refactors Zephyr's CMake build system
    - More modular, readable, documentation
 * **Removes hardcoded assumptions of GNU toolchain**
    - A necessary step towards Zephyr supporting commercial compilers
 * Generalizes/lifts unsharable toolchain usage out to separate "API" macros
 * Macros exist for abstracting the compiler, **linker and binary utilities**
 * API macro names are **intent-based**; they don't make assumptions of low-level details. Details are for the particular toolchain to worry about -- all that matters is that these toolchains perform the same _logical_ task, albeit in very different ways
 * Preserves existing CMake containers/data structures so a common framework can still be shared among disparate toolchains


As discussed in the TSC. See presentation: 

- Personal (open): https://drive.google.com/file/d/1hGgk9ITFTXS3QoOr3GTFlC46VhEopT0Y/view
- Official (access controlled):  https://drive.google.com/file/d/0B5NybbZNGEEdZWpxZzFmLVY3N3hxMzBHOUNQaVM4LTg2Sjd3/view

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon commented:
Rebasing this to make CI run, will take some time.

mped-oticon commented:
The link from TSC MoM is behind access control for some reason.

Here is a working (personal) copy:
https://drive.google.com/file/d/1hGgk9ITFTXS3QoOr3GTFlC46VhEopT0Y/view?usp=sharing

mped-oticon commented on cmake/compiler/api_gcc_freestanding.cmake:
```diff
@@ -0,0 +1,50 @@
+    -Wno-main
+    ${NOSTDINC_F}
+    ${TOOLCHAIN_C_FLAGS}
+  )
```
This is still a mix-bag. Will be cleaned up.

mped-oticon committed:
cmake: document toolchain abstraction in-place

* ${ZEPHYR_TOOLCHAIN_VARIANT}.cmake
* ${COMPILER}.cmake
* ${LINKER}.cmake
* ${BINTOOLS}.cmake
* toolchain_cc
* toolchain_cc_preprocess_linker_pass
* toolchain_ld
* offsets
* usermode
* toolchain_ld_link_step_0_init
* toolchain_ld_link_step_1_base
* toolchain_ld_link_step_2_generate

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

kapedev commented on cmake/app/boilerplate.cmake:
```diff
@@ -25,91 +22,131 @@
+  message(FATAL_ERROR "Source directory equals build directory.\
+ In-source builds are not supported.\
+ Please specify a build directory, e.g. cmake -Bbuild -H.")
+endif()
```
The -B switch is not supported for CMake < 3.13 ( https://cmake.org/cmake/help/v3.13/manual/cmake.1.html ) The current minimum requirement is Cmake >= 3.8.2. This will likely bump to 3.10.2 soon

mped-oticon replied:
But it does work for older CMake too, they are just undocumented: https://stackoverflow.com/a/11144109/275297

SebastianBoe commented on cmake/toolchain/zephyr.cmake:
```diff
@@ -27,7 +28,7 @@ elseif("${ARCH}" STREQUAL "xtensa")
   set(TOOLCHAIN_INCLUDES
     ${SYSROOT_DIR}/include/arch/include
     ${SYSROOT_DIR}/include
-    )
```
It would be helpful from a review perspective, and from a PR maintainability perspective, if the patches that are desirable (even without applying the toolchain abstraction patches) could be upstreamed before the PR is ready for review.

mped-oticon replied:
The PR has now been split across multiple commits. That takes out much of the noise from things like whitespace changes.
I think it is ready for review.

ruuddw commented:
I like the refactoring Mark! Still some work on the 'Hmmm' sections for Userspace support. Any thoughts on handling differences in linker command files formats (esp. the generated parts)?

mped-oticon commented:
> I like the refactoring Mark! 

Thank you! 

> Still some work on the 'Hmmm' sections for Userspace support. 

To be pragmatic, I will defer that work until we need it: I bet the first many commercial-compilers are targetting systems without MPU and User mode threading.

> Any thoughts on handling differences in linker command files formats (esp. the generated parts)?

Yes. The generated linker script is really just a preprocessed version of `linker.ld` which can come from the Board, SoC or via `CONFIG_HAVE_CUSTOM_LINKER_SCRIPT`. Every `linker.ld` in Zephyr is meant for GNU ld compatible linkers, so that leaves us with `CONFIG_HAVE_CUSTOM_LINKER_SCRIPT` where we, for now, have to supply a custom-written linker file through. Then it will pass through the preprocessor like the others, which is fine.

mped-oticon committed:
cmake: adapt 'support for images without gaps' for abstracted toolchains

This is the rebased version of:

  c086b935aa kconfig: cmake: Support for images without gaps.

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon committed:
cmake: cosmetics: improve wording and fix typos

Collect preliminary improvements:
 * Wording
 * Commentary
 * Spacing
 * Fix typos
 * TODO

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon committed:
cmake: boilerplate: move compiler-flag checks into boilerplate

CheckCCompilerFlag and CheckCXXCompilerFlag does not belong to
extensions.cmake since they are standard CMake files. Boilerplate
seems more appropriate.

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon committed:
cmake: document CMAKE_SYSTEM_-variables and CMAKE_FIND_ROOT_PATH

Aid user in understanding what CMake expects of:

 * CMAKE_SYSTEM_NAME
 * CMAKE_SYSTEM_PROCESSOR
 * CMAKE_SYSTEM_VERSION
 * CMAKE_FIND_ROOT_PATH_MODE_PROGRAM
 * CMAKE_FIND_ROOT_PATH_MODE_LIBRARY
 * CMAKE_FIND_ROOT_PATH_MODE_INCLUDE
 * CMP0002

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon committed:
cmake: boilerplate: Spin out finding git+python to separate cmake files

Follow the same convention as that of ccache.
Improve readability of boilerplate.cmake file.

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon committed:
cmake: boilerplate: reorganize into headlines

* Collect related things
* Introduce headlines for overview's sake

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon committed:
cmake: boilerplate: move 'app'-library to be with its documentation

Declare the zephyr app library earlier and document it then and there.

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon committed:
cmake: minor comment fixes of MPU, C99

* MPU comments
* C99 comment on same line
* toolchain_cc_preprocess_linker_pass

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon committed:
cmake: move post-link commands: copy-exe and check-link-map are common

Not all post-link commands should be abstracted out. We can retain
commonality because the copy and checking of map file is under Kconfig.

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon committed:
cmake: move CMakeLists.txt's documentation-header to top

Inform user early on that /CMakeLists.txt is not the top-level file.

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon committed:
cmake: extensions: use ZEPHYR_DEFINE_FLAG_C and CMAKE_INCLUDE_FLAG_C

While many toolchains do, not *all*, understand -D and -I flags as
Define and Include flags.

CMake partially abstracts this already, but only for Include flags via
CMAKE_INCLUDE_FLAG_C and CMAKE_INCLUDE_FLAG_CXX.

We introduce ZEPHYR_DEFINE_FLAG_C and ZEPHYR_DEFINE_FLAG_CXX
to correct for the lack of Define flag abstraction in CMake.

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

mped-oticon committed:
cmake: toolchain abstraction: LINKER, BINTOOLS and toolchain api

Lift usage of GNU (gcc,ld,objcopy,..) out into intent-based macros, to
ease porting efforts towards new toolchains. This was presented to the
TSC on Oct 17 2018 "Supporting non-GNU toolchains" and approved during
Edinburgh F2F.

An example of one of these intent-based macros is toolchain_cc(). It is
a macro which:

 * Is compiler-specific. Hence all compilers supported by Zephyr must
   implement this macro.
 * Has an intent: Expands to all compiler options that Zephyr wants but
   which are not portable across compilers/toolchains.
 * Must implement Zephyr's implicit expectations. E.g. freestanding.
 * Must implement Zephyr's explicit Kconfig expectations. E.g. warning
   and optimization levels.

Taken together, this forms an "API" or "contract" on toolchain_cc().
This API is documented in the follow-up commit, at CMake call sites in
CMakeLists.txt to keep it in one place.


This commit reorganizes toolchain.cmake:

 * Introduce LINKER, following same pattern of COMPILER
 * Introduce BINTOOLS, following same pattern of COMPILER
 * Introduce headlines for overview's sake
 * Add host toolchain for POSIX
 * Remove special case for POSIX, no longer needed

And refactors CMakeLists.txt:
 * Introduce headlines for overview's sake
 * Introduce toolchain_cc()
 * Introduce toolchain_cc_preprocess_linker_pass()
 * Introduce toolchain_ld()
 * Introduce toolchain_ld_link_step_0_init()
 * Introduce toolchain_ld_link_step_1_base()
 * Introduce toolchain_ld_link_step_2_generated()
 * Introduce toolchain_bintools_generate_isr()
 * Introduce toolchain_bintools_post_link()
 * Introduce toolchain_bintools_print_size()
 * Introduce toolchain_bintools_usermode()


The flow is very much retained, but now just extended a bit:

hello_world/CMakeLists.txt:   include(app/boilerplate.cmake)

  app/boilerplate.cmake:        include(extensions.cmake)
  app/boilerplate.cmake:        include(toolchain.cmake)
    toolchain.cmake:              include(toolchain/zephyr.cmake)
      toolchain/zephyr.cmake:       set(COMPILER   gcc)
      toolchain/zephyr.cmake:       set(LINKER     ld)
      toolchain/zephyr.cmake:       set(BINTOOLS   binutils)

    toolchain.cmake:              include(compiler/gcc.cmake)
      compiler/gcc.cmake:           find_program(CMAKE_C_COMPILER...)
      compiler/gcc.cmake:           include(compiler/api_gcc.cmake)

    toolchain.cmake:              include(linker/ld.cmake)
      linker/ld.cmake:              find_program(CMAKE_LINKER...)
      linker/ld.cmake:              include(linker/api_ld.cmake)

    toolchain.cmake:              include(bintools/binutils.cmake)
      bintools/binutils.cmake:      find_program(CMAKE_OBJCOPY...)
      bintools/binutils.cmake:      find_program(CMAKE_OBJDUMP...)
      bintools/binutils.cmake:      include(bintools/api_binutils.cmake)

  CMakeLists.txt:               toolchain_cc() #defined in api_gcc.cmake

Signed-off-by: Mark Ruvald Pedersen <mped@oticon.com>

SebastianBoe reviewed (changes requested):
NACK'ing to allow time for review.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -56,297 +63,80 @@ zephyr_include_directories(
-  ${USERINCLUDE}
-  ${STDINCLUDE}
+  ${USERINCLUDE}  # unused
+  ${STDINCLUDE}   # unused
```
Let's either drop these, or document them as user-settable.

I'd suggest dropping them.

I'd also suggest doing this in a separate PR.

mped-oticon replied:
@SebastianBoe agreed, I can do that in a follow-up PR.

SebastianBoe replied:
In general it is better to do it in a preceeding PR, as that keeps PR's small. This is especially true for big PR's like this one.

aescolar reviewed (commented):
I just did a quick review and only found a couple of nit picks.

SebastianBoe commented on cmake/compiler/api_llvm.cmake:
```diff
@@ -0,0 +1,34 @@
+include(${TOOLCHAIN_ROOT}/cmake/compiler/api_gcc_security.cmake)
+include(${TOOLCHAIN_ROOT}/cmake/compiler/api_gcc_asm.cmake)
+include(${TOOLCHAIN_ROOT}/cmake/compiler/api_gcc_cpp.cmake)
+include(${TOOLCHAIN_ROOT}/cmake/compiler/api_gcc_warnings.cmake)
```
${CMAKE_CURRENT_LIST_DIR}/api_gcc_asm.cmake

(or CMAKE_CURRENT_SOURCE_DIR, I can never remember ...)

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -56,297 +63,80 @@ zephyr_include_directories(
-    )
-endif()
+# Discover if the default C standard has been overridden
+# So far, only POSIX-based boards does this
```
> So far, only POSIX-based boards does this

This comment is not likely to be updated, and should therefore be dropped, which is unfortunate, because it is useful information.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -56,297 +63,80 @@ zephyr_include_directories(
+##
+##     1. Toolchain-specific and Zephyr-agnostic:
+##          Covers: ABI, call-convention, code generation options, etc.
+##          Where: Defines and includes should be should be captured by
```
should be should be

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -56,297 +63,80 @@ zephyr_include_directories(
+##          Where: Defines and includes should be should be captured by
+##            {TOOLCHAIN_C_FLAGS, TOOLCHAIN_INCLUDES} set in ${COMPILER}.cmake.
+##            Compiler options that are not defines or includes, should be set
+##            directly in ${COMPILER}.cmake or ${LINKER}.cmake.
```
I'm not able to follow what the "Where" section is trying to say here.

mped-oticon replied:
It means the physical location of these options.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -56,297 +63,80 @@ zephyr_include_directories(
+
+
+## @Brief: toolchain_cc: Macro that expands into compiler options.
+## @Brief: toolchain_ld: Macro that expands into linker options.
```
"expands into" sounds like it expands into a string of options.

Perhaps

Macro that sets compiler options.

would be clearer.

mped-oticon replied:
Agreed.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -56,297 +63,80 @@ zephyr_include_directories(
+##            toolchain_cc down into toolchain_cc_{security,optimizations,cpp,
+##            freestanding,warnings,diagnostics,asm}. Subdividing like this can
+##            improve reuse (e.g. Clang accepts the gcc-names of C++ standards),
+##            and demarcate responsibility.
```
"demarcate"

It is a part of the Zephyr coding standard to use simple english.

mped-oticon replied:
OK, will fix.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -56,297 +63,80 @@ zephyr_include_directories(
+##     4. Toolchain-agnostic and Zephyr-specific:
+##          Covers: Common defines (__ZEPHYR__, BUILD_VERSION), common
+##            include paths (kernel/include, ...), etc.
+##          Where: CMakeLists.txt and most of Zephyr tree (e.g. subsys,drivers).
```
> CMakeLists.txt

It's ambiguous whether we are referring to zephyr/CMakeLists.txt or generally CMakeLists.txt files.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -376,6 +166,8 @@ endif()
 configure_file(version.h.in ${PROJECT_BINARY_DIR}/include/generated/version.h)
 
+
+
```
Too much whitespace ...

SebastianBoe commented:
This PR is way too big and does too many things, it is too difficult to review it, and too difficult to maintain it until it is ready for merging.

You need to first do a refactor/documentation PR, and then, after it is merged, another PR that introduces the feature you need.

NACKing this as it violates the Contribution guidelines:

> One general practice we encourage, is to make small, controlled changes. This practice simplifies review, makes merging and rebasing easier, and keeps the change history clear and clean.

https://docs.zephyrproject.org/latest/contribute/contribute_guidelines.html#contribution-workflow

Examples:
https://github.com/zephyrproject-rtos/zephyr/commit/a3ee546f7c73d377167605a3752d945858646fb5#diff-af3b638bc2a3e6c650974192a53c7291R82
https://github.com/zephyrproject-rtos/zephyr/commit/a3ee546f7c73d377167605a3752d945858646fb5#diff-32dd6f2b8d30fbe1624076ef9d1e4049R42

SebastianBoe commented:
I am missing documentation for the 'intents'. E.g. I would expect "security" to be documented. Or maybe even finer granularity than "security". But I definitely expect documentation at a finer level than toolchain_cc().

mped-oticon commented:
> I am missing documentation for the 'intents'. 

That documentation has actually been provided. E.g. `toolchain_cc()`:

```
## @Brief: toolchain_cc: Macro that expands into compiler options.
## @Details:
##   toolchain_cc must have been defined by compiler/${COMPILER}.cmake, where
##     COMPILER was set by toolchain/${ZEPHYR_TOOLCHAIN_VARIANT}.cmake.
##
##  --> 2. Toolchain-specific and Zephyr-specific:
##          Covers:
##            a. Mapping Kconfig "intents" to toolchain-specific flags (e.g. W,
##              CONFIG_SPEED_OPTIMIZATIONS, CONFIG_READABLE_ASM).
##            b. Mapping Zephyr expectations to toolchain-specific flags
##              PIC/PIE, freestanding,  inclusion of autoconf.h, expected
##              overflow behavior, etc.
##          Where: Should be placed in toolchain_cc() or toolchain_ld().
##          Note: It may be neat, but optional, to further subdivide
##            toolchain_cc down into toolchain_cc_{security,optimizations,cpp,
##            freestanding,warnings,diagnostics,asm}. Subdividing like this can
##            improve reuse
```


> E.g. I would expect "security" to be documented. Or maybe even finer granularity than "security". But I definitely expect documentation at a finer level than toolchain_cc().

Well that documentation is by definition toolchain-specific. Hence I can only document the intention of what that macro should do, which was done above.

`toolchain_cc_security()` is a subdivision for GCC, as also written above. This subdivision can get documentation attached to it (which I will do), but is entirely specific to GCC. 

To be clear: `toolchain_cc_security()` is therefore not a requirement to implement by every toolchain. But a toolchain may do so, if it sees fit.

mped-oticon commented:
> This PR is way too big and does too many things, it is too difficult to review it, and too difficult to maintain it until it is ready for merging.

I feel the commits do what they say, and are described well enough.

The whole point of this is to clean up Zephyr's usage of CMake which is frankly already unreadable and dare I say, unmaintainable. 

**Of course there is going to be a gigantic diff when hundreds of lines of hard-codedness is replaced with a single `toolchain_cc()`.** 

Regarding contribution workflow, it is not the law.
Big/revolutionary changes should be limited yes, but it isn't always possible is it?

```
12f8f76165 - Sebastian Bøe : Introduce cmake-based rewrite of KBuild
644 files changed, 7815 insertions(+)
```

Compared to my biggest commit:
```
34 files changed, 1102 insertions(+), 845 deletions(-)
```

SebastianBoe commented:
> To be clear: toolchain_cc_security() is therefore not a requirement to implement by every toolchain. But a toolchain may do so, if it sees fit.

There is something that is not quite right here. Let us take this line for instance;

zephyr_compile_options_ifdef(CONFIG_STACK_CANARIES -fstack-protector-all)

When STACK_CANARIES is enabled. It is not up to the toolchain to determine whether guard variables should be used or not. The user has enabled the feature, and expects it to be present no matter which toolchain he uses. If the toolchain does not support guard variables, this must be restricted through Kconfig.

SebastianBoe commented:
> Big/revolutionary changes should be limited yes, but it isn't always possible is it?

Not always, but it is in this instance. I have to insist that the contribution guidelines are followed before I invest time into getting this reviewed.

mped-oticon commented:
> Not always, but it is in this instance. I have to insist that the contribution guidelines are followed before I invest time into getting this reviewed.

Alright, I understand. Many small PRs it is.
To avoid further NACKs, what are your max-size requirements?

SebastianBoe commented:
> To avoid further NACKs, what are your max-size requirements?

There is no LOC limit. The goal is for the code to be reviewable, traceable, and bisectable.

These types of things should be not be in the same commit (unless required for bisectability):

Adding docs/comments to existing functionality.
Refactoring existing functionality.
Changing existing functionality.
Adding new functionality.

Notably, these types of things can/should be in the same commit:

Adding new functionality and adding docs/comments to this new functionality.
Changing functionality X and changing functionality Y, when X is a backwards-incompatible change and requires changing Y.

It is okay to group together many unrelated comments into the same commit. Because adding a comment is unlikely to require a revert or a helpful commit message. But refactorings should not be grouped past a handful of recatorings because a refactoring might accidentally not be a refactoring, and might require a revert or a bisect.

Functionality should be separated as far as possible into independent features in individual commits. But only when it serves the purpose of being: reviewable, traceable and bisectable.

mped-oticon commented:
Will split up across smaller PRs.
